HMCL/src/main/java/org/jackhuang/hmcl/util/FileSaver.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Hello Minecraft! Launcher
+ * Copyright (C) 2025 huangyuhui <huanghongxun2008@126.com> and contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package org.jackhuang.hmcl.util;
+
+import org.jackhuang.hmcl.task.Schedulers;
+import org.jackhuang.hmcl.util.io.FileUtils;
+
+import java.nio.file.Path;
+import java.util.*;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static org.jackhuang.hmcl.util.logging.Logger.LOG;
+
+/**
+ * @author Glavo
+ */
+public final class FileSaver extends Thread {
+
+    private static final Pair<Path, String> SHUTDOWN = Pair.pair(null, null);
+
+    private static final BlockingQueue<Pair<Path, String>> queue = new LinkedBlockingQueue<>();
+
+    private static volatile boolean running = false;
+    private static final ReentrantLock runningLock = new ReentrantLock();
+    private static volatile boolean installedShutdownHook;
+
+    private static void doSave(Map<Path, String> map) {
+        for (Map.Entry<Path, String> entry : map.entrySet()) {
+            Path file = entry.getKey();
+            LOG.info("Saving file " + file);
+            try {
+                FileUtils.saveSafely(file, entry.getValue());
+            } catch (Throwable e) {
+                LOG.warning("Failed to save " + file, e);
+            }
+        }
+    }
+
+    private static void onExit() {
+        shutdown();
+        runningLock.lock();
+        try {
+            HashMap<Path, String> map = new HashMap<>();
+            for (Pair<Path, String> pair : queue) {
+                if (pair != SHUTDOWN)
+                    map.put(pair.getKey(), pair.getValue());
+            }
+            doSave(map);
+        } finally {
+            runningLock.unlock();
+        }
+    }
+
+    public static void save(Path file, String content) {
```

Comment:
Assume `running` is false and no FileSaver is started yet.
1. Caller thread A & B call `save()` simultaneously. Both of them pass the `!running` check (at L77) and start two new threads C & D (at L78).
2. Thread C acquires the lock, starts a FileSaver thread E, and then finishes the task.
3. Thread E acquires the lock and executes the loop in `run()`, holding the lock until shutdown.
4. Now thread D is stuck at L79 because the lock is held by thread E forever.